src/Control/SimpleResourceURLGenerator.php

```diff
@@ -65,40 +67,103 @@ public function setNonceStyle($nonceStyle)
+                    ManifestFileFinder::RESOURCES_DIR,
+                    substr($relativePath, strlen(ManifestFileFinder::VENDOR_DIR))
+                );
+            }
         } else {
-            // Use normal string
-            $absolutePath = preg_replace('/\?.*/', '', Director::baseFolder() . '/' . $relativePath);
-            $exists = file_exists($absolutePath);
+            // Remove querystring args, normalise path
+            if (strpos($relativePath, '?') !== false) {
+                list($relativePath, $query) = explode('?', $relativePath);
+            }
+            $relativePath = Filesystem::normalisePath($relativePath, true);
+
+            // Detect public-only request
+            $withPublic = stripos($relativePath, Director::publicDir() . DIRECTORY_SEPARATOR) === 0;
+            if ($withPublic) {
+                $relativePath = substr($relativePath, strlen(Director::publicDir() . DIRECTORY_SEPARATOR));
+            }
+            $absolutePath = null;
+            $exists = false;
+            if (!Director::publicDir()) {
+                // @todo Non-public dir support will be removed in 5.0, so remove this block there
+                // Throw if public path, but no public dir
+                if ($withPublic) {
+                    trigger_error('Requesting a public resource without a public folder has no effect', E_USER_WARNING);
+                }
+                $absolutePath = Filesystem::joinPaths(Director::baseFolder(), $relativePath);
+                $exists = file_exists($absolutePath);
+
+                // Rewrite vendor/ to resources/ folder
+                if (stripos($relativePath, ManifestFileFinder::VENDOR_DIR . DIRECTORY_SEPARATOR) === 0) {
+                    $relativePath = Filesystem::joinPaths(
+                        ManifestFileFinder::RESOURCES_DIR,
+                        substr($relativePath, strlen(ManifestFileFinder::VENDOR_DIR))
+                    );
+                }
+            } else {
+                // Search public folder first, and unless `public/` is prefixed, also private base path
+                $publicPath = Filesystem::joinPaths(Director::publicFolder(), $relativePath);
+                $privatePath = Filesystem::joinPaths(Director::baseFolder(), $relativePath);
+                if (file_exists($publicPath)) {
+                    // String is a literal url comitted directly to public folder
+                    $absolutePath = $publicPath;
+                    $exists = true;
+                } elseif (!$withPublic && file_exists($privatePath)) {
+                    // String is private but exposed to resources/
+                    $absolutePath = $privatePath;
+                    $exists = true;
+                    $relativePath = Filesystem::joinPaths(ManifestFileFinder::RESOURCES_DIR, $relativePath);
+                }
+            }
         }
         if (!$exists) {
             trigger_error("File {$relativePath} does not exist", E_USER_NOTICE);
```

Comment:
If I request an “exposed” file, but I’ve forgotten to run `composer vendor-expose`, is this the error I’d see? Do you think there’d be any value in adding a hint about running that command here?

Reply:
No, that would occur if you generated a url for a file that didn't exist at all.